contracts/0.8.25/vaults/VaultFactory.sol

```diff
@@ -0,0 +1,109 @@
+
+    function setManagementFee(uint256 _newManagementFee) external;
+
+    function setPerformanceFee(uint256 _newPerformanceFee) external;
+
+    function grantRole(bytes32 role, address account) external;
+
+    function revokeRole(bytes32 role, address account) external;
+}
+
+contract VaultFactory is UpgradeableBeacon {
+    address public immutable delegationImpl;
+
+    /// @param _owner The address of the VaultFactory owner
+    /// @param _stakingVaultImpl The address of the StakingVault implementation
+    /// @param _delegationImpl The address of the Delegation implementation
+    constructor(
+        address _owner,
+        address _stakingVaultImpl,
+        address _delegationImpl
+    ) UpgradeableBeacon(_stakingVaultImpl, _owner) {
+        if (_delegationImpl == address(0)) revert ZeroArgument("_delegation");
+
+        delegationImpl = _delegationImpl;
+    }
+
+    /// @notice Creates a new StakingVault and Delegation contracts
+    /// @param _stakingVaultParams The params of vault initialization
+    /// @param _initializationParams The params of vault initialization
+    function createVault(
+        bytes calldata _stakingVaultParams,
+        IDelegation.InitializationParams calldata _initializationParams,
+        address _lidoAgent
+    ) external returns (IStakingVault vault, IDelegation delegation) {
+        if (_initializationParams.manager == address(0)) revert ZeroArgument("manager");
+        if (_initializationParams.operator == address(0)) revert ZeroArgument("operator");
+
+        vault = IStakingVault(address(new BeaconProxy(address(this), "")));
+
+        delegation = IDelegation(Clones.clone(delegationImpl));
+
+        delegation.initialize(address(this), address(vault));
+
+        delegation.grantRole(delegation.LIDO_DAO_ROLE(), _lidoAgent);
+        delegation.grantRole(delegation.MANAGER_ROLE(), _initializationParams.manager);
+        delegation.grantRole(delegation.OPERATOR_ROLE(), _initializationParams.operator);
+        delegation.grantRole(delegation.DEFAULT_ADMIN_ROLE(), msg.sender);
+
+        delegation.grantRole(delegation.OPERATOR_ROLE(), address(this));
+        delegation.grantRole(delegation.MANAGER_ROLE(), address(this));
+        delegation.setManagementFee(_initializationParams.managementFee);
+        delegation.setPerformanceFee(_initializationParams.performanceFee);
+
+        //revoke roles from factory
+        delegation.revokeRole(delegation.MANAGER_ROLE(), address(this));
+        delegation.revokeRole(delegation.OPERATOR_ROLE(), address(this));
+        delegation.revokeRole(delegation.DEFAULT_ADMIN_ROLE(), address(this));
+        delegation.revokeRole(delegation.LIDO_DAO_ROLE(), address(this));
+
+        vault.initialize(address(delegation), _stakingVaultParams);
+
+        emit VaultCreated(address(delegation), address(vault));
+        emit DelegationCreated(msg.sender, address(delegation));
+    }
```

Comment:
:brain: role setup has dependencies on the `Delegation` internals which is not perfect